crates/core/src/nats/client.rs

```diff
@@ -0,0 +1,223 @@
+use std::{sync::Arc, time::Duration};
+
+use bytes::Bytes;
+use tracing::info;
+
+use super::{
+    types::{
+        JetStreamConfig,
+        JetStreamContext,
+        NatsConsumer,
+        NatsStream,
+        PullConsumerConfig,
+    },
+    NatsError,
+    Subject,
+};
+use crate::types::BoxedResult;
+
+#[derive(Debug, Clone)]
+pub struct NatsClient {
+    pub url: String,
+    pub conn_id: String,
+    pub conn: Arc<async_nats::Client>,
+    pub jetstream: Arc<JetStreamContext>,
+}
+
+impl NatsClient {
+    pub async fn connect(
+        url: &str,
+        conn_id: &str,
+        nkey: Option<String>,
+    ) -> Result<Self, NatsError> {
+        let conn_id = conn_id.to_string();
+        let conn = Self::create_conn(url, nkey).await?;
+        let context =
+            Arc::new(async_nats::jetstream::new(conn.as_ref().clone()));
+
+        info!("Connected to NATS server at {}", url);
+        Ok(Self {
+            url: url.to_string(),
+            conn_id,
+            conn,
+            jetstream: context,
+        })
+    }
+
+    async fn create_conn(
+        url: &str,
+        nkey: Option<String>,
+    ) -> Result<Arc<async_nats::Client>, NatsError> {
+        let options = async_nats::ConnectOptions::new()
+            .connection_timeout(Duration::from_secs(30))
+            .max_reconnects(10);
+
+        if nkey.is_none() {
+            return Err(NatsError::NkeyNotProvided {
+                url: url.to_owned(),
+            })
+        }
+
+        let conn =
+            async_nats::connect_with_options(&url, options.nkey(nkey.unwrap()))
+                .await
+                .map_err(|e| NatsError::ConnectError {
+                    url: url.to_owned(),
+                    source: e,
+                })?;
+
+        Ok(Arc::new(conn))
```

Comment:
Do we need the `Arc`?

```suggestion
        Ok(conn)
```

Reply:
I mentioning in the previews comment related to this why

Reply:
`Arc` is usually used as a wrapper type. Returning it from a factory function doesn't seem right. The caller should get to decide if they want to wrap it with an `Arc`, `Rc`, or `Box`

Reply:
I tried again without `Arc` and this time it worked, maybe the bug was something not related to the pointer itself when I tried earlier